### PR TITLE
Reuse connection

### PR DESCRIPTION
**Pull request by panzerdev**
- Establish connection to server only once
- Reuse the once created connection
- Close connection after usage

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -21,12 +21,16 @@ const (
 	ClientKey  = "cert/client.key"
 )
 
+type Dailer func(target string) (*grpc.ClientConn, error)
```
Comment (bbengfort) [inline client.go:24]: Minor typo: Dailer --> Dialer and requires comment if exported.

File: client.go [hunk 2]
```diff
@@ -21,12 +21,16 @@ const (
...
 type PingClient struct {
-	Name     string
-	Delay    time.Duration
-	Limit    uint
-	sequence int64
+	Name       string
+	Delay      time.Duration
+	Limit      uint
+	sequence   int64
+	Connection *grpc.ClientConn
```
Comment (bbengfort) [inline client.go:32]: Is there any reason to make the connection public?

File: client.go [hunk 3]
```diff
@@ -53,13 +57,13 @@ func (c *PingClient) Run(addr string) error {
 			return nil
 		}
 
-		pong, err := c.Ping(addr)
+		pong, err := c.Echo(context.Background(), c.Next())
 		if err != nil {
-			return err
+			return fmt.Errorf("failed echo RPC call: %s", err)
```
Comment (bbengfort) [inline client.go:62]: Much better error message, thanks!

File: client.go [hunk 4]
```diff
@@ -113,9 +117,43 @@ func (c *PingClient) PingMutualTLS(addr string) (*pb.Pong, error) {
...
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to %s: %s", addr, err)
+	}
+	return conn, nil
+}
+
 // PingTLS is a helper method for server-side encryption that does not expect
 // client authentication or credentials. It is mostly here for benchmarking.
-func (c *PingClient) PingTLS(addr string) (*pb.Pong, error) {
+func TLS(addr string) (*grpc.ClientConn, error) {
```
Comment (bbengfort) [inline client.go:156]: I think I'd prefer this to be renamed ServerTLS to make it more even with MutualTLS

File: sping.go [hunk 5]
```diff
@@ -14,12 +16,18 @@ func init() {
 }
 
 // NewClient returns a ping client with the specified options.
-func NewClient(name string, delay int64, limit uint) *PingClient {
+func NewClient(dailer Dailer, address, name string, delay int64, limit uint) *PingClient {
```
Comment (bbengfort) [inline sping.go:19]: Ok, so your proposed interface is that you pass a dialer to the `NewClient` function and the dialers are defined in the code. Any reason this is preferred over making the dialers methods of the client?

**Review by bbengfort — APPROVED** [review]
@panzerdev thanks for the update and connection reuse! I'll go ahead and merge it in; the comments were more for me if I wanted to make changes in the future, but feel free to respond to any of them.

File: client.go (hunk 1 above)
Comment (panzerdev) [reply client.go:24]: Damn... I always mix that up. Deamon, Dialer... Thanks for noticing :smile:

File: client.go (hunk 2 above)
Comment (panzerdev) [reply client.go:32]: One has to be able to call `Close()` on that Connection. There could be a better solution by returning a `func` to defer when returning a new `Client`.

File: client.go (hunk 3 above)
Comment (panzerdev) [reply client.go:62]: I just moved that from somewhere else.. Can't really take credit for that :wink:

File: client.go (hunk 4 above)
Comment (panzerdev) [reply client.go:156]: There is also die Option to add `Dialer` to those `func`s to give them names like `DialTLS` or `DialInsecure` or so?

Comment (bbengfort) [reply client.go:156]: Yeh, I like that idea!

File: sping.go (hunk 5 above)
Comment (panzerdev) [reply sping.go:19]: I didn't want to change too much so I had to have a way of getting a `Connection` with different setup for the `Client`. As I like the idea to have a `type MyAwesomeFunc func() ` and then pass that in I went for this approach. I think there might bit better ones, but in this case it fulfills the need to be able to create different `Connections` and have a `Client` use that.

**Pull request merged**